### ISSUE TITLE
Proposal: add IList, IСollection and IEnumerable support to Span, ReadOnlySpan, Memory and ReadOnlyMemory

### ISSUE THREAD

**Issue by AlexRadch** [dotnet/corefxlab#2463]
Adding new types (Span, ReadOnlySpan, Memory and ReadOnlyMemory) in dotNet increased the speed of work with memory buffers and significantly reduced memory consumption.

Due to the location of the Span and the ReadOnlySpan structures on the stack, they can not implement interfaces. And accordingly Span and ReadOnlySpan structures do not implement such useful interfaces as IList and etc.

Also, the Memory and the ReadOnlyMemory structures do not implement these interfaces because their implementation will be very performance inefficient. https://github.com/dotnet/coreclr/issues/19787

This state of affairs has a significant drawback, since many standard and useful algorithms are written for IList, ICollection and IEnumerable interfaces and can not now be used when working with new structures. Such algorithms should be duplicated for new structures. Also now if you create a new class, then there are situations that you must implement the same algorithm twice. One for standard interfaces for collections, and one for new buffer structures.

I propose to extend dotNet in such a way as to add implementation of interfaces for structures located on the stack. Then structures such as Span and ReadOnlySpan will support IList, ICollection and IEnumerable interfaces. Also, Memory and ReadOnlyMemory structures will support the fast implementation of IList, ICollection and IEnumerable interfaces.

I suggest adding in dotNet the concept of the location of the instance (interface, class or structure). There are only two types of location: the heap and the stack. And add in dotNet control of the correct location of instances at compile time.

For example, now if we declare the following basic interfaces with location support:
```cs
//    public interface IEnumerator<out T> as before

    public interface IEnumerable`A<out T> // where A can be stackalloc or heapalloc
    {
        // the result of GetEnumerator() method (IEnumerator<T>) have the same location as IEnumerable<T> interface
        IEnumerator<T>`A GetEnumerator(); 
    }

    public interface ICollection`A<T> : IEnumerable`A<T>
    // IEnumerable<T> have the same location as ICollection<T>
    {
        // as before
    }

    public interface IList`A<T> : ICollection`A<T>, IEnumerable`A<T>
    // ICollection`A<T> and IEnumerable<T> have the same location as IList<T>
    {
        // as before
    }
```

Then we can add support for such interfaces to Span and ReadOnlySpan structures:
```cs
    public ref struct Span<T>: IList`stackalloc<T> 
    // Span structure implement only stack allocated IList<T> interface
    {
    }
```

Also now we can add a quick implementation of these interfaces for Memory and ReadOnlyMemory structures
```cs
    public struct Memory<T>: IList`stackalloc<T>
    // the Memory structure can fast implement only stack allocated IList<T> interface
    {
    }
```

Such interfaces can not be stored in a heap or in an object field, so can not be used in any method. But can be used in methods that support stack allocated interfaces and classes. Such methods can not store such interfaces and classes on the heap.

For example, we can add such support to Linq:
```cs
    public static class Enumerable
    {
        // Aggregate does not store IEnumerable<TSource> source on the heap
        public static TSource Aggregate`A<TSource>(this IEnumerable`A<TSource> source, Func<TSource, TSource, TSource> func);

        // Concat can store first and second parameters in a result instance that have the same location as first and second parameters
        public static IEnumerable`A<TSource> Concat`A<TSource>(this IEnumerable`A<TSource> first, IEnumerable`A<TSource> second);

        // etc.
    }
```

Also for interfaces, classes and structures allocated on stack, the compiler must ensure that the order of destruction is correct.

**Comment by YohDeadfall** [comment]
This would make it possible to pass a reference to the interface which is implemented by a structure. Therefore you can store the reference somewhere on the heap. To prevent such things the CLR must check that the reference doesn't point to the heap and thow an exception if this happens, and this will drastically decrease performance.

**Comment by AlexRadch** [comment]
> This would make it possible to pass a reference to the interface which is implemented by a structure. Therefore you can store the reference somewhere on the heap. To prevent such things the CLR must check that the reference doesn't point to the heap and throw an exception if this happens, and this will drastically decrease performance.

IList interface will be implemented by Span structure. Can you add example of decrease performance?

**Comment by GrabYourPitchforks** [comment]
Per https://github.com/dotnet/csharplang/issues/1833, there's little drive from the language side of things to add interface support to ref structs.

Given that interfaces aren't likely to happen, is there something else we can do in the Framework side of things to make this better? Any Span-based extension methods we're missing that can help unblock your immediate scenarios?

**Comment by YohDeadfall** [comment]
I don't think that something should be added due to the idea of `Span<T>`. It was designed to increase performance, but wrappers around it would lead to bad results.

By the way, there is a similar issue, dotnet/corefx#5360.

**Comment by KrzysztofCwalina** [comment]
I would rather develop a set of adapters that adapt Memory\<T\> to IList\<T\>, than implement IList\<T\> on Memory<T>. Such interface implementations might be useful short term, but after algorithms nativelly understand spans, it will be much more perfromant to get span out of Memory\<T\> once, and then call as many manipulation algorithms as possible.

**Comment by highfield** [comment]
Is really this thread still open?
I read it just now, and today I've posted a related question on SO: https://stackoverflow.com/questions/57567192/is-there-any-common-interface-between-readonlyspant-and-ireadonlylistt
I must admit that the question on SO is badly written (shame on me), but I still couldn't understand why a way to work better with arrays it cannot be used against tons of already-written functions for that. At least, I hope the sample code on the post would clarify where the point is.
@KrzysztofCwalina could you elaborate your suggestion? I also would avoid interfaces in favor of adapters or something similar.
